File: go-backend-interview-prep/08_http_and_backend/01_http_server_basics/main.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net"
	"net/http"
	"time"
)

func helloHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "text/plain")
	fmt.Fprintln(w, "Hello from Go HTTP server!")
}

func healthHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]string{
		"status": "ok",
		"time":   time.Now().Format(time.RFC3339),
	})
}

func main() {
	mux := http.NewServeMux()
	mux.HandleFunc("/hello", helloHandler)
	mux.HandleFunc("/health", healthHandler)

	// Use a random port for demo; in production use a fixed port
	listener, err := net.Listen("tcp", "127.0.0.1:8080")
	if err != nil {
		// Port 8080 busy -- pick random
		listener, err = net.Listen("tcp", "127.0.0.1:0")
		if err != nil {
			log.Fatal(err)
		}
	}

	addr := listener.Addr().String()
	fmt.Printf("Server running on http://%s\n", addr)
	fmt.Println("  GET /hello  -> greeting")
	fmt.Println("  GET /health -> JSON health check")

	// Auto-shutdown after 30s for demo purposes
	server := &http.Server{Handler: mux}
	go func() {
		time.Sleep(30 * time.Second)
		fmt.Println("\nAuto-shutting down (demo timeout)...")
		server.Close()
	}()

	if err := server.Serve(listener); err != http.ErrServerClosed {
		log.Fatal(err)
	}
}

File: go-backend-interview-prep/08_http_and_backend/01_http_server_basics/main.py
"""HTTP server basics -- Python equivalent of the Go example.

Uses only the stdlib http.server module (no Flask/FastAPI).
"""

import json
import threading
import time
import urllib.request
from http.server import HTTPServer, BaseHTTPRequestHandler


class MyHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == "/hello":
            self.send_response(200)
            self.send_header("Content-Type", "text/plain")
            self.end_headers()
            self.wfile.write(b"Hello from Python HTTP server!\n")

        elif self.path == "/health":
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            body = json.dumps({
                "status": "ok",
                "time": time.strftime("%Y-%m-%dT%H:%M:%S%z"),
            })
            self.wfile.write(body.encode())

        else:
            self.send_response(404)
            self.end_headers()
            self.wfile.write(b"not found\n")

    def log_message(self, *args):
        pass  # silence default stderr logging


def main() -> None:
    server = HTTPServer(("127.0.0.1", 0), MyHandler)
    host, port = server.server_address
    base = f"http://{host}:{port}"
    print(f"Server running on {base}")
    print("  GET /hello  -> greeting")
    print("  GET /health -> JSON health check")

    # Run server in background thread
    threading.Thread(target=server.serve_forever, daemon=True).start()

    # --- Demo: hit the endpoints ---
    print("\n--- Demo requests ---")

    resp1 = urllib.request.urlopen(f"{base}/hello", timeout=5)
    print(f"GET /hello  -> {resp1.status}: {resp1.read().decode().strip()}")

    resp2 = urllib.request.urlopen(f"{base}/health", timeout=5)
    data = json.loads(resp2.read().decode())
    print(f"GET /health -> {resp2.status}: {data}")

    server.shutdown()
    print("\nServer stopped.")


if __name__ == "__main__":
    main()
